features/storage/kvstore/tdbstore/TDBStore.cpp

```diff
@@ -0,0 +1,1421 @@
+    return ret;
+}
+
+int TDBStore::iterator_open(iterator_t *it, const char *prefix)
+{
+    key_iterator_handle_t *handle;
+
+    if (!_is_initialized) {
+        return MBED_ERROR_NOT_READY;
+    }
+
+    if (!it) {
+        return MBED_ERROR_INVALID_ARGUMENT;
+    }
+
+    _mutex.lock();
+
+    handle = new key_iterator_handle_t;
+    *it = reinterpret_cast<iterator_t>(handle);
+
+    if (prefix && strcmp(prefix, "")) {
+        handle->prefix = new char[strlen(prefix) + 1];
+        strcpy(handle->prefix, prefix);
+    } else {
+        handle->prefix = 0;
+    }
+    handle->ram_table_ind = 0;
+
+    _mutex.unlock();
+
+    return MBED_SUCCESS;
+}
+
+int TDBStore::iterator_next(iterator_t it, char *key, size_t key_size)
+{
+    ram_table_entry_t *ram_table = (ram_table_entry_t *) _ram_table;
+    key_iterator_handle_t *handle;
+    int ret;
+    uint32_t actual_data_size, hash, flags, next_offset;
+
+    if (!_is_initialized) {
+        return MBED_ERROR_NOT_READY;
+    }
+
+    _mutex.lock();
+
+    handle = reinterpret_cast<key_iterator_handle_t *>(it);
+
+    ret = MBED_ERROR_ITEM_NOT_FOUND;
+
+    while (ret && (handle->ram_table_ind < _num_keys)) {
+        ret = read_record(_active_area, ram_table[handle->ram_table_ind].bd_offset, _key_buf,
+                          0, 0, actual_data_size, 0, true, false, false, false, hash, flags, next_offset);
+        if (ret) {
+            goto end;
+        }
+        if (!handle->prefix || (strstr(_key_buf, handle->prefix) == _key_buf)) {
+            if (strlen(_key_buf) >= key_size) {
+                ret = MBED_ERROR_INVALID_SIZE;
+                goto end;
+            }
+            strcpy(key, _key_buf);
+        } else {
+            ret = MBED_ERROR_ITEM_NOT_FOUND;
+        }
+        handle->ram_table_ind++;
+    }
+
+end:
+    _mutex.unlock();
+    return ret;
+}
+
+int TDBStore::iterator_close(iterator_t it)
+{
+    key_iterator_handle_t *handle;
+
+    if (!_is_initialized) {
+        return MBED_ERROR_NOT_READY;
+    }
+
+    _mutex.lock();
+
+    handle = reinterpret_cast<key_iterator_handle_t *>(it);
+    delete[] handle->prefix;
+    delete handle;
+
+    _mutex.unlock();
+
+    return MBED_SUCCESS;
+}
+
+int TDBStore::reserved_data_set(const void *reserved_data, size_t reserved_data_buf_size)
+{
+    reserved_trailer_t trailer;
+    int os_ret, ret = MBED_SUCCESS;
+
+    if (reserved_data_buf_size > RESERVED_AREA_SIZE) {
+        return MBED_ERROR_INVALID_SIZE;
+    }
+
+    _mutex.lock();
+
+    ret = do_reserved_data_get(0, RESERVED_AREA_SIZE);
+    if ((ret == MBED_SUCCESS) || (ret == MBED_ERROR_INVALID_DATA_DETECTED)) {
+        ret = MBED_ERROR_WRITE_FAILED;
+        goto end;
+    } else if (ret != MBED_ERROR_ITEM_NOT_FOUND) {
+        goto end;
+    }
+
+    ret = write_area(_active_area, 0, reserved_data_buf_size, reserved_data);
+    if (ret) {
+        goto end;
+    }
+
+    trailer.trailer_size = sizeof(trailer);
+    trailer.data_size = reserved_data_buf_size;
```

Comment:
Minor note: Does data_size make sense here? Wouldn't you need the data_size to find the trailer in the first place?

I suspect this needs both a prefix (size) and a suffix (crc).

Reply:
data size is part of the *trailer*, placed at a fixed offset after the data.

Reply:
Then why store data_size at all if it's already known?

Reply:
It's not known. The trailer is placed at the maximum size offset of the reserved data space, but the size can be smaller. If a smaller size has been stored, I would like to know it.